### DISCUSSION TITLE
delaying-strongly-typed-options-configuration-using-postconfigure-in-asp-net-core/

### DISCUSSION THREAD

**Discussion by andrewlock** [andrewlock/blog-comments#277] [Announcements]
# delaying-strongly-typed-options-configuration-using-postconfigure-in-asp-net-core/

In this post I show how to use PostConfigure() with strongly-typed options to run configuration actions after all other Configure() actions are applied.

https://andrewlock.net/delaying-strongly-typed-options-configuration-using-postconfigure-in-asp-net-core/

[a comment by an automated account is not shown]

**Comment by AlexSapple** [comment]
Thanks for this article! - just one question/can of worms

Is `PostConfigure()` only intended to be used for validation or setting default values? - MS docs suggest that but..

what if you had some extension method that "does stuff" - I'm thinking of an extension method that sets up some authentication protection on an API - so the extension method may want to do things like build Authorization policies and execute things like `services.AddAuthentication()` and basically do a load of stuff based on the given options. Thoughts...?

    // Default library options are overridden
    // by bound configuration values.
    services.Configure(options);
    services.PostConfigure<MyOptionsType>(optionsPostConfigure =>
    {
        //validation stuff - great :)
        if(string.IsNullOrWhiteSpace(optionsPostConfigure?.Authority) 
           || optionsPostConfigure.Audiences.Any(a => !string.IsNullOrWhiteSpace(a)))
             throw new ArgumentException(nameof(options));

        //but is it legit to use this place to do more stuff
        services.AddAuthentication(JwtBearerDefaults.AuthenticationScheme).AddJwtBearerOptions => { /*loads of stuff in here*/
    });
